include/libpmemobj++/experimental/vector.hpp

```diff
@@ -0,0 +1,399 @@
+void
+vector<T>::_dealloc()
+{
+	assert(pmemobj_tx_stage() == TX_STAGE_WORK);
+
+	if (_data != nullptr) {
+		_destruct_at_end(0);
+		if (pmemobj_tx_free(*_data.raw_ptr()) != 0)
+			throw transaction_free_error(
+				"failed to delete persistent memory "
+				"object");
+		_capacity = 0;
+	}
+}
+
+/**
+ * Private helper function. Must be called during transaction. Assumes that
+ * there is enough space for additional elements. Copy constructs elements at
+ * the end of underlying array based on given parameters.
+ *
+ * @param[in] count number of elements to construct.
+ * @param[in] value value of all constructed elements.
+ *
+ * @throw rethrows constructor exception.
+ *
+ * @pre pmemobj_tx_stage() == TX_STAGE_WORK
+ * @pre if initialized, range [end(), end() + count) must be snapshotted in
+ * current transaction
+ * @pre _capacity >= count + _size
+ * @pre value is valid argument for value_type copy constructor
+ *
+ * @post _size = _size + count
+ */
+template <typename T>
+void
+vector<T>::_construct_at_end(size_type count, const value_type &value)
+{
+	assert(pmemobj_tx_stage() == TX_STAGE_WORK);
+	assert(_capacity >= count + _size);
+
+	pointer dest = _data.get() + static_cast<difference_type>(_size);
+	pointer end = dest + count;
+	for (; dest != end; ++dest)
+		detail::create<value_type, const value_type &>(dest, value);
+	_size += count;
+}
+
+/**
+ * Private helper function. Must be called during transaction. Assumes that
+ * there is enough space for additional elements and input arguments satisfies
+ * InputIterator requirements. Constructs elements in underlying array with the
+ * contents of the range [first, last). The first and last arguments must
+ * satisfy InputIterator requirements. This overload only participates in
+ * overload resolution if InputIt satisfies InputIterator.
+ *
+ * @param[in] first first iterator.
+ * @param[in] last last iterator.
+ *
+ * @throw rethrows constructor exception.
+ *
+ * @pre pmemobj_tx_stage() == TX_STAGE_WORK
+ * @pre if initialized, range [end(), end() + std::distance(first, last)) must
+ * be snapshotted in current transaction
+ * @pre _capacity >= std::distance(first, last) + _size
+ * @pre InputIt is InputIterator
+ * @pre InputIt::reference is valid argument for value_type copy constructor
+ *
+ * @post _size = _size + std::distance(first, last)
+ */
+template <typename T>
+template <typename InputIt,
+	  typename std::enable_if<
+		  detail::is_input_iterator<InputIt>::value &&
+			  std::is_constructible<
+				  T,
+				  typename std::iterator_traits<
+					  InputIt>::reference>::value,
+		  InputIt>::type *>
+void
+vector<T>::_construct_at_end(InputIt first, InputIt last)
+{
+	assert(pmemobj_tx_stage() == TX_STAGE_WORK);
+	assert(_capacity >=
+	       static_cast<size_type>(std::distance(first, last)) + _size);
+
+	pointer dest = _data.get() + static_cast<difference_type>(_size);
+	_size += static_cast<size_type>(std::distance(first, last));
+	for (; first != last; ++dest)
```

Comment:
if you'll move "++" to loop body this can be replaced with while loop.
Please do that or move the other ++ from loop body to this for.